stripped/colabs/GrupowanieEM_clean_output_stripped.py

```diff
@@ -0,0 +1,507 @@
+import datashader.transfer_functions as tf
+import colorcet as cc
+import holoviews as hv
+import hvplot.pandas
+from holoviews.operation.datashader import datashade
+import geoviews as gv
+import geoviews.tile_sources as gts
+from holoviews import opts
+from IPython.display import IFrame
+from IPython.core.display import display
+from bokeh.plotting import show, output_notebook
+
+nasa2020 = spark.sql(''' SELECT * FROM nasa_full''').where(nasa_full.Year == 2020).drop('Year')
+nasa2020.show(5)
+
+nasa_an = preprocessed_loader(spark, "/content/drive/MyDrive/BigMess/NASA/NASA_an.csv")
+
+nasa_an = nasa_an.withColumnRenamed("lon", "lon_sam").withColumnRenamed("lat", "lat_sam")
+nasa_an.show(5)
+
+NASA_2020_an = (
+     nasa2020
+     .join(
+        nasa_an,
+        [
+             nasa2020.lon==nasa_an.lon_sam ,
+             nasa2020.lat==nasa_an.lat_sam
+         ],
+          "inner"
+      )
+      .drop('lat_sam','lon_sam')
+  )
+
+NASA_2020_an.show(5)
+
+NASA_2020_an_pd = NASA_2020_an.toPandas()
+
+NASA_2020_an_pd.head(5)
+
+#agreggating selected features' monthly data over a year:
+
+# a) calculating annual averages:
+NASA2020_annual_means = (
+                         NASA_2020_an_pd[['lon', 'lat', 'Evap','PotEvap','RootMoist','Rainf','SoilM_100_200cm', 'GVEG']]
+                         .groupby( by=['lon', 'lat']
+                                  ).mean()
+                        ).reset_index()
+
+# b) calculating the total annual rainfall (sum):
+NASA2020_rainfall_sum = (
+                         NASA_2020_an_pd[['lon', 'lat', 'Rainf']]
+                         .groupby( by=['lon', 'lat']
+                                  ).sum()
+                        .rename(columns={"Rainf": "Annual Rainfall"})
+                        .reset_index()
+                        )
+
+# c) calculating annual medians for selected features:
+NASA2020_medians = (
+                   NASA_2020_an_pd[['lon', 'lat', 'PotEvap', 'Evap', 'SoilM_100_200cm', 'AvgSurfT', 'SoilT_40_100cm' ]]
+                   .groupby( by=['lon', 'lat']
+                            ).median()
+                   .rename(columns={'PotEvap': "PotEvap_Median", 'Evap': 'Evap_Median', 'AvgSurfT': 'AvgSurfT_Median',
+                                    'SoilT_40_100cm': "SoilT40_100_Median", 'SoilM_100_200cm': 'SoilM_100_200cm_Median'}
+                          )
+                   .reset_index()
+                   )
+
+# Merging into one dataset:
+NASA2020_annual_anotated = NASA2020_annual_means.merge(NASA2020_medians, how='inner', on=['lon', 'lat'])
+
+
+NASA2020_annual_anotated = NASA2020_annual_anotated.merge( NASA2020_rainfall_sum, how='inner', on=['lon', 'lat'])
+
+NASA2020_annual_anotated.head()
+
+def assert_add_column(dataset: pd.DataFrame, feature_name: str, month: int) -> None:
+  assert (1<= month <= 12), f'Invalid month number'
+  assert feature_name in dataset.columns, f'The attribute "{feature_name}" is not a column in the DataFrame (full_dataset)'
```

Comment:
Takich funkcji się nie robi. Normalnie dodaj te asercje do funkcji

Reply:
Wzorowałam się na Panu - Pan zrobił osobną funkcję do asercji w moim notatniku ze zbiorami treningowymi i testowymi :wink: Ale dobrze, zmienię to.